File: leetcode/2108/c/main.c
#include <stdbool.h>
#include <string.h>

bool isPalindrome(char* word) {
    int i = 0, j = strlen(word) - 1;

    while (i < j) {
        if (word[i++] != word[j--]) {
            return false;
        }
    }
    return true;
}

char* firstPalindrome(char** words, int wordsSize) {
    for (int i = 0; i < wordsSize; i++) {
        if (isPalindrome(words[i])){
            return words[i];
        }
    }

    return "";
}

File: leetcode/2108/rs/src/main.rs
pub struct Solution;

impl Solution {
    pub fn first_palindrome(words: Vec<String>) -> String {
        let found = words.iter().find(|&word| Self::is_palindrome(word));
        match found {
            Some(str) => str.to_owned(),
            None => "".to_string(),
        }
    }

    fn is_palindrome(string: &str) -> bool {
        let mid = string.len().div_euclid(2);
        for (a, b) in string.as_bytes()[0..mid]
            .iter()
            .zip(string.as_bytes()[mid..].iter().rev())
        {
            if *a != *b {
                return false;
            }
        }

        true
    }
}

fn main() {
    println!("Hello, world!");
}
